### ISSUE TITLE
Don't require sheet name for references inside named expressions

### ISSUE THREAD

**Issue by sequba** [handsontable/hyperformula#1206]
### Description

As @thilgen wrote in #1191 

> Currently [Named Expressions](https://hyperformula.handsontable.com/guide/named-expressions.html) require reference formulas to use **absolute cell references**
> 
> Hyperformula, Excel, and Google Sheets all allow absolute references to be defined like this (includes the sheet-decorator)
> ```
> Sheet1!$A$1
> Sheet1!$B$1:$D$2
> Sheet1!$2:$2
> Sheet1!$B:$B
> ```
> However, Excel and Google Sheets **also** allow for absolute references to be defined **without** a sheet-decorator (and the absolute references will be automatically scoped to the current sheet)
> ```
> $A$1
> $B$1:$D$2
> $2:$2
> $B:$B
> ```
> **This PR provides support for auto-scoping absolute references within a named reference formula to the sheet where the named reference is installed**
> 
> ### Example Data
> 
> ```
> const engine = HyperFormula.buildFromSheets({
>   'Sheet1': [[]],
>   'Sheet2': [[]],
> })
> ```
> 
> | Expression | Evaluated As |
> | :----------  | :----------- |
> | `engine.addNamedExpression('foo', '=$B$1', 0)` |   `=Sheet1!$B$1`  | 
> | `engine.addNamedExpression('foo', '=CONCATENATE($A$1,".2")', 1)` | `=CONCATENATE(Sheet2!$A$1,".2")`  |
> | `engine.addNamedExpression('foo', '=SUM($B$1,$D$1)', 1)` | `=SUM(Sheet2!$B$1,Sheet2!$D$1)`  |
> | `engine.addNamedExpression('foo', '=SUM($B$1,Sheet1!$D$1)', 1)` | `=SUM(Sheet2!$B$1,Sheet1!$D$1)`  |
> | `engine.addNamedExpression('foo', '=SUM($B$1:$D$2)', 1)` | `=SUM(Sheet2!$B$1:$D$2)`|
> | `engine.addNamedExpression('foo', '=SUM($2:$2)', 0)` | `=SUM(Sheet1!$2:$2)`|
> | `engine.addNamedExpression('foo', '=SUM($B:$B)', 1)` | `=SUM(Sheet2!$B:$B)`  |

**Comment by sequba** [comment]
We need to verify compliance with ODFF and compatibility with other popular spreadsheet software.

**Comment by thilgen** [comment]
# Note: I think my proposed change is incorrect

## Why I proposed this enhancement:

* I have an application that has several tables - each representing a different account ledger (e.g. Checking, Savings, etc.)
  * You can see a pared-down example in this JSFiddle ([hyperformula-auto-scoped-named-references](https://jsfiddle.net/thilgen/e5v3m6cL/1/))
* These tables each have a column where an `account id` can be entered
* And another column that uses `VLOOKUP` to get the name of an account that matches the `account id` from a `company_accounts` table

**checking_ledger**
| account id | account name (read-only)                                                                                 | date      | description | amount |
| :--------- | :------------------------------------------------------------------------ | :-------- | :---------- | :----- |
| 14         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 2/15/2021 | public storage | 75.00  |
| 10          | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 3/15/2021 | lawn care   | 45.00  |
| 10         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 4/15/2021 | lawn care   | 45.00  |
| 12         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 5/15/2021 | supplies    | 12.50  |

**savings_ledger**
| account id | account name (read-only)                                                  | date      | description | amount |
| :--------- | :------------------------------------------------------------------------ | :-------- | :---------- | :----- |
| 10         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 2/15/2021 | door repair | 75.00  |
| 9          | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 3/15/2021 | new roof  | 45.00  |
| 10         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 4/15/2021 | lawn care   | 45.00  |
| 14         | `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")` | 5/15/2021 | office space  | 12.50  |

**company_accounts**
| account id | account name |
| :--------- | :------------------------ |
| 12 | "Office Supplies"            |
| 9   | "Capital Improvements" |
| 10 | "Maintenance"             | 
| 14 | "Rent"       |

## Utilizing Named Expressions

* Obviously repeating that same formula expression over and over and over again is pretty cumbersome - and is begging for a Named Expression
* However, because Named Expressions require sheet-decorated absolute addresses, the best I could do would be to define a named Expression for each sheet

```
/* must explicitly use checking_ledger! decoration */
engine.addNamedExpression( 
  'GetCompanyForChecking', 
  `=IFNA(VLOOKUP(INDEX(checking_ledger!$A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")`
  0
)

/* must explicitly use savings_ledger! decoration */
engine.addNamedExpression( 
  'GetCompanyForSavings, 
  `=IFNA(VLOOKUP(INDEX(savings_ledger!$A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")`
  0
)
```
## So that's a little better...
* But still not really what I am hoping to do
* Which is to register a single Named Expression at the global scope - use it in all sheets - and have non-sheet-decorated absolute references auto-scope when they are evaluated **(which is definitely NOT what I originally proposed)**
```
engine.addNamedExpression( 
  'GetCompany', 
  `=IFNA(VLOOKUP(INDEX($A:$A, ROW()),company_accounts!A:B,2,FALSE()),"")`
)
```
**checking_ledger**
| account id | account name (read-only)                                                                                 | date      | description | amount |
| :--------- | :------------------------------------------------------------------------ | :-------- | :---------- | :----- |
| 14         | `=GetCompany` | 2/15/2021 | public storage | 75.00  |
| 10          | `=GetCompany` | 3/15/2021 | lawn care   | 45.00  |
| 10         | `=GetCompany` | 4/15/2021 | lawn care   | 45.00  |
| 12         | `=GetCompany` | 5/15/2021 | supplies    | 12.50  |

**savings_ledger**
| account id | account name (read-only)                                                  | date      | description | amount |
| :--------- | :------------------------------------------------------------------------ | :-------- | :---------- | :----- |
| 10         | `=GetCompany` | 2/15/2021 | door repair | 75.00  |
| 9          | `=GetCompany` | 3/15/2021 | new roof  | 45.00  |
| 10         | `=GetCompany` | 4/15/2021 | lawn care   | 45.00  |
| 14         | `=GetCompany` | 5/15/2021 | office space  | 12.50  |

**Comment by thilgen** [comment]
Some more info from the Hyperformula docs about absolute reference that lack sheet-decorators

https://hyperformula.handsontable.com/guide/cell-references.html#relative-references

![image](https://user-images.githubusercontent.com/9309340/227667510-faebc59b-7fc0-46ef-b99a-632cdc27bcdb.png)

**Comment by sequba** [comment]
As far as I can tell, pull request https://github.com/handsontable/hyperformula/pull/1191 introduces a purely syntactic change. Sheet decorator (`Sheet1!`) is not required, but references inside named expressions are still evaluated as absolute (bound to the sheet specified by the `sheetId` in third parameter of [addNamedExpression](https://hyperformula.handsontable.com/api/classes/hyperformula.html#named-expressions)).

@thilgen Please, correct me if I am wrong: What you really wanted, is the ability to use references with relative sheet inside named expressions, right?

**Comment by thilgen** [comment]
Yeah that was mostly it - but when I realized that what I really wanted was to define the Named Expression at the global scope - I realized that what I really wanted was for the auto-scoping to take place when the Named Expression is actually inserted into the dependency tree for a sheet. And that was really quite different than the approach I was taking. I am not giving up on this - but I think the approach of applying the sheet id during named expression registration isn't quite right.
